applications/StructuralMechanicsApplication/custom_processes/metrics_spr_error_process.cpp

```diff
@@ -0,0 +1,659 @@
+        if(mSetElementNumber == true)
+            new_element_size *= std::sqrt((std::pow(rEnergyNormOverall, 2) + std::pow(rErrorOverall, 2))/mElementNumber) * mTargetError;
+        else
+            new_element_size *= std::sqrt((rEnergyNormOverall*rEnergyNormOverall+rErrorOverall*rErrorOverall)/mThisModelPart.Elements().size())*mTargetError;
+
+
+        // Check if element sizes are in specified limits. If not, set them to the limit case
+        if(new_element_size < mMinSize)
+            new_element_size = mMinSize;
+        if(new_element_size > mMaxSize)
+            new_element_size = mMaxSize;
+
+        it_elem->SetValue(ELEMENT_H, new_element_size);
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<SizeType TDim>
+void SPRMetricProcess<TDim>::CalculateMetric()
+{
+    const bool meshing_application_compiled = KratosComponents<Variable<Vector>>::Has("MMG_METRIC");
+    KRATOS_WARNING_IF("CalculateMetric", !meshing_application_compiled) << "Please compile the MeshingApplication in order to compute the metric" << std::endl;
+
+    // Actually computing the metric
+    if (meshing_application_compiled) {
+        // Getting metric variable
+        const Variable<Vector>& metric_variable = KratosComponents<Variable<Vector>>::Get("MMG_METRIC");
+
+        // Iteration over all nodes
+        NodesArrayType& nodes_array = mThisModelPart.Nodes();
+        const int num_nodes = static_cast<int>(nodes_array.size());
+
+        #pragma omp parallel for
+        for(int i_node = 0; i_node < num_nodes; ++i_node) {
+            auto it_node = nodes_array.begin() + i_node;
+            /**************************************************************************
+            ** Determine nodal element size h:
+            ** if mAverageNodalH == true : the nodal element size is averaged from the element size of neighboring elements
+            ** if mAverageNodalH == false: the nodal element size is the minimum element size from neighboring elements
+            */
+            double h_min = 0.0;
+            auto& neigh_elements = it_node->GetValue(NEIGHBOUR_ELEMENTS);
+            for(auto i_neighbour_elements = neigh_elements.begin(); i_neighbour_elements != neigh_elements.end(); i_neighbour_elements++){
+                const double element_h = i_neighbour_elements->GetValue(ELEMENT_H);
+                if(mAverageNodalH == false) {
+                    if(h_min == 0.0 || h_min > element_h)
+                        h_min = element_h;
+                } else {
+                    h_min += element_h;
+                }
+            }
+
+            if(mAverageNodalH == true)
+                h_min = h_min/static_cast<double>(neigh_elements.size());
+
+            // Set metric
+            BoundedMatrix<double, TDim, TDim> metric_matrix = ZeroMatrix(TDim, TDim);
```

Comment:
This code block could be put in another function to set the metric and called here.

Reply:
actually i would say that the calculation of the error and the calculation of the metrics belong to two completely different files... one may very well be interested in the error without being interested in the metrics